packages/chevrotain/src/parse/parser/traits/recognizer_engine.ts

```diff
@@ -210,35 +213,30 @@ export class RecognizerEngine {
     this.shortRuleNameToFull[shortName] = ruleName
     this.fullRuleNameToShort[ruleName] = shortName
 
-    function invokeRuleWithTry(this: MixedInParser, args: any[]) {
+    function invokeRuleWithTry(this: MixedInParser, ...args: ARGS[]): R {
       try {
+        this.ruleInvocationStateUpdate(shortName, ruleName, this.subruleIdx)
+        this.subruleIdx = 0
         if (this.outputCst === true) {
           impl.apply(this, args)
           const cst = this.CST_STACK[this.CST_STACK.length - 1]
           this.cstPostRule(cst)
-          return cst
+          return cst as unknown as R
         } else {
           return impl.apply(this, args)
         }
       } catch (e) {
-        return this.invokeRuleCatch(e, resyncEnabled, recoveryValueFunc)
+        return this.invokeRuleCatch(e, resyncEnabled, recoveryValueFunc) as R
       } finally {
         this.ruleFinallyStateUpdate()
       }
     }
 
-    const wrappedGrammarRule = function (
-      this: MixedInParser,
-      idxInCallingRule: number = 0,
```

Comment:
getting rid of the default assignment here is good, it may improve performance slightly